### PR TITLE
[JENKINS-52729] Launcher.ProcStarter.stdout(TaskListener) discards remotability

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-52729](https://issues.jenkins-ci.org/browse/JENKINS-52729).

I suspect there is also a lower-priority encoding bug here, but I am not bothering with that for now.

### Proposed changelog entries

* Developer: `Launcher.ProcStarter.stdout(TaskListener)` did not properly send its argument over a Remoting channel to an agent.

### CODE REVIEW

**Comment by jglick** [comment]
Sigh, apparently `echo` is not an actual command on Windows, but built into `cmd.exe`? Presumably so you can `@echo off`. Whereas on Linux it is both. Will fix test accordingly.

**Comment by jglick** [comment]
FTR [this call](https://github.com/jenkinsci/jenkins/blob/1a11fa39b567fd7ff64e61fab23aa18a123f2d97/core/src/main/java/hudson/tasks/CommandInterpreter.java#L109) seems to be what would be used by typical freestyle build steps.

File: core/src/main/java/hudson/Launcher.java [hunk 1]
```diff
@@ -295,7 +297,9 @@ public ProcStarter stdout(@CheckForNull OutputStream out) {
          * @return {@code this}
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: It makes sense to explicitly set the requirement that the class is remotable in Javadoc

**Review by oleg-nenashev — APPROVED** [review]
Yes, it would be nice to get it fixed in the context of External Logging stories. In #3557 reference impls I work it around by creating my own Launcher, but such approach would be much better

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
By the way, `stederr()` has identical problem. I suggest fixing it here as well so that we do not get into surprises with some implementations really using 2 streams

**Comment by oleg-nenashev** [comment]
OTOH we could create a follow-up task for stederr

File: core/src/main/java/hudson/Launcher.java (hunk 1 above)
Comment (jglick) [reply core/src/main/java/hudson/Launcher.java]: All `TaskListener` implementations are expected to be remotable but I will clarify this in Javadoc.

**Comment by jglick** [comment]
Test failure is a common flake:

```
java.lang.AssertionError: deletion stopped
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at hudson.UtilTest.testDeleteRecursive_onWindows(UtilTest.java:489)
```

**Review by oleg-nenashev — APPROVED** [review]
OK, let's merge it as is. Public API does not change here, so we can fix STDERR in a follow-up if I discover the usa-case for it

**Pull request merged**